app/models/concerns/password_validator.rb

```diff
@@ -2,18 +2,62 @@
 module PasswordValidator
   extend ActiveSupport::Concern
 
+  MAX_PREVIOUS_PASSWORD = 3
+  PASSWORD_VALIDITY_MONTHS = 6
+
+  def password_needs_update?
+    (password_updated_on + PASSWORD_VALIDITY_MONTHS.months) < Time.zone.now
+  end
+
   private
 
-  def password_complexity
-    return if good_password?
+  def password_updated_on
+    password_histories.last.created_at
+  end
 
-    error_message = I18n.t('devise.registrations.password_complexity_error')
-    errors.add(:password, error_message)
+  def create_password_history
+    return unless saved_change_to_encrypted_password?
+
+    password_histories.create(encrypted_password: encrypted_password)
+
+    return if password_histories.count <= MAX_PREVIOUS_PASSWORD
+
+    password_histories.first.destroy
   end
 
-  def good_password?
+  def password_complexity
+    google_oauth2_enabled? ||
+      password.blank? ||
+      (matches_format? && not_in_used_passwords?)
+  end
+
+  def matches_format?
     password_regex =
       /^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).*$/
-    google_oauth2_enabled? || password.blank? || (password =~ password_regex)
+    return true if password =~ password_regex
+
+    errors.add(:password, I18n.t('devise.registrations.password_errors.format'))
+    false
+  end
+
+  def not_in_used_passwords?
+    password_histories.pluck(:encrypted_password).each do |ep|
```

Comment:
Just for clarity, let's rename `ep` to `password`

Reply:
I think using password will conflict with already existing password attr in user, will use old_password.

Reply:
What about using encrypted_password instead?

Reply:
There is a filed with same name in the user model, even though the encrypted_password will be local to the loop, I think will be good if avoid it, what do you think.

Reply:
Updated it to encrypted_password.